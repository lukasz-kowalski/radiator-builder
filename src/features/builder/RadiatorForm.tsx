"use client";

import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { Input, Select } from "@/components/form";
import { Button } from "@/components/general";
import { RadiatorFamily } from "@/api/dto/radiator";
import { radiatorFormSchema } from "@/features/builder/schema/RadiatorForm.schema";
import { Inputs } from "@/features/builder/types/inputs";

interface Props {
  families: RadiatorFamily[];
  onSubmit: (data: Inputs) => void;
}

export default function RadiatorForm({ families, onSubmit }: Props) {
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<Inputs>({
    defaultValues: {
      radiatorFamily: "",
      radiatorLengthFrom: 0,
    },
    mode: "onBlur",
    resolver: zodResolver(radiatorFormSchema),
  });

  return (
    <form aria-label="form" onSubmit={handleSubmit(onSubmit)}>
      <div className="flex flex-col gap-1 sm:gap-4 sm:flex-row sm:items-end">
        <Select
          {...register("radiatorFamily")}
          aria-label="Radiator family"
          defaultValue=""
          label="Radiator family"
          id="radiatorFamily"
          error={errors.radiatorFamily?.message}
          options={families.map((item) => ({
            label: item.name,
            value: item.id,
          }))}
        />

        <Input
          label="Radiator length from (mm)"
          error={errors.radiatorLengthFrom?.message}
          {...register("radiatorLengthFrom")}
          id="radiatorLengthFrom"
          type="number"
        />
        <Input
          label="Radiator length to (mm)"
          error={errors.radiatorLengthTo?.message}
          {...register("radiatorLengthTo")}
          id="radiatorLengthTo"
          type="number"
        />
      </div>

      <div className="flex justify-end gap-4">
        <Button onClick={() => reset()}>Clear</Button>
        <Button
          type="submit"
          variant="confirm"
          aria-label="Search for radiators"
        >
          Search
        </Button>
      </div>
    </form>
  );
}
