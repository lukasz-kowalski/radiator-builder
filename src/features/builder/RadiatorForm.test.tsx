import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import RadiatorForm from "@/features/builder/RadiatorForm";

const familiesMock = [
  { id: "1", name: "Family 1" },
  { id: "2", name: "Family 2" },
];

describe("RadiatorForm", () => {
  it("should render the RadiatorForm with input fields", () => {
    render(<RadiatorForm families={familiesMock} onSubmit={jest.fn()} />);

    const selectElement = screen.getByRole("combobox");

    expect(selectElement).toBeInTheDocument();
    expect(
      screen.getByLabelText("Radiator length from (mm)")
    ).toBeInTheDocument();
    expect(
      screen.getByLabelText("Radiator length to (mm)")
    ).toBeInTheDocument();
    expect(screen.getByText("Search")).toBeInTheDocument();
    expect(screen.getByText("Clear")).toBeInTheDocument();
  });

  it("should display error messages when values of length fields are too high", async () => {
    const onSubmitMock = jest.fn();

    render(<RadiatorForm families={familiesMock} onSubmit={onSubmitMock} />);

    const inputLengthFrom = screen.getByLabelText("Radiator length from (mm)");
    const inputLengthTo = screen.getByLabelText("Radiator length to (mm)");

    await userEvent.type(inputLengthFrom, "10001");
    await userEvent.tab();
    await userEvent.type(inputLengthTo, "20000");
    await userEvent.tab();

    const errorMessages = await screen.findAllByText(
      "Length cannot be greater than 10000"
    );
    expect(errorMessages).toHaveLength(2);
  });

  it("should call onSubmit with correct data when form is submitted", async () => {
    const onSubmitMock = jest.fn();

    render(<RadiatorForm families={familiesMock} onSubmit={onSubmitMock} />);

    const familySelect = screen.getByLabelText("Radiator family");
    const inputLengthFrom = screen.getByLabelText("Radiator length from (mm)");
    const inputLengthTo = screen.getByLabelText("Radiator length to (mm)");

    await userEvent.selectOptions(familySelect, "1");
    await userEvent.type(inputLengthFrom, "300");
    await userEvent.type(inputLengthTo, "600");

    await userEvent.click(screen.getByText("Search"));

    await waitFor(() => {
      const [data] = onSubmitMock.mock.calls[0];
      expect(data).toEqual({
        radiatorFamily: "1",
        radiatorLengthFrom: 300,
        radiatorLengthTo: 600,
      });
    });
  });

  it("should clear the form when Clear button is clicked", async () => {
    const onSubmitMock = jest.fn();

    render(<RadiatorForm families={familiesMock} onSubmit={onSubmitMock} />);

    const familySelect = screen.getByLabelText("Radiator family");
    const inputLengthFrom = screen.getByLabelText("Radiator length from (mm)");
    const inputLengthTo = screen.getByLabelText("Radiator length to (mm)");

    await userEvent.selectOptions(familySelect, "1");
    await userEvent.type(inputLengthFrom, "300");
    await userEvent.type(inputLengthTo, "600");

    expect(screen.getByLabelText("Radiator family")).toHaveValue("1");
    expect(screen.getByLabelText("Radiator length from (mm)")).toHaveValue(300);
    expect(screen.getByLabelText("Radiator length to (mm)")).toHaveValue(600);

    const clearButton = screen.getByText("Clear");
    await userEvent.click(clearButton);

    await waitFor(async () => {
      expect(screen.getByLabelText("Radiator family")).toHaveValue("");
      expect(screen.getByLabelText("Radiator length from (mm)")).toHaveValue(0);
      expect(screen.getByLabelText("Radiator length to (mm)")).toHaveValue(
        null
      );
    });

    expect(onSubmitMock).not.toHaveBeenCalled();
  });

  it("should match the snapshot", () => {
    const onSubmitMock = jest.fn();

    const { asFragment } = render(
      <RadiatorForm families={familiesMock} onSubmit={onSubmitMock} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
