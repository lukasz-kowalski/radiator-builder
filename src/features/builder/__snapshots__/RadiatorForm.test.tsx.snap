// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadiatorForm should match the snapshot 1`] = `
<DocumentFragment>
  <form
    aria-label="form"
  >
    <div
      class="flex flex-col gap-1 sm:gap-4 sm:flex-row sm:items-end"
    >
      <div
        class="flex flex-col w-full"
      >
        <div
          class="h-[20px]"
        />
        <select
          aria-errormessage="error-radiatorFamily"
          aria-invalid="false"
          aria-label="Radiator family"
          class="p-1 border-2 h-10 bg-white"
          id="radiatorFamily"
          name="radiatorFamily"
        >
          <option
            disabled=""
            hidden=""
            selected=""
            value=""
          >
            Radiator family
          </option>
          <option
            value="1"
          >
            Family 1
          </option>
          <option
            value="2"
          >
            Family 2
          </option>
        </select>
        <div
          class="h-[3rem]"
        >
          <p
            class="text-red-600 text-sm"
            id="error-radiatorFamily"
          />
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <label
          class="text-sm"
          for="radiatorLengthFrom"
        >
          Radiator length from (mm)
        </label>
        <input
          aria-errormessage="error-radiatorLengthFrom"
          aria-invalid="false"
          class="p-2 border-2 h-10 bg-white"
          id="radiatorLengthFrom"
          name="radiatorLengthFrom"
          type="number"
        />
        <div
          class="h-[3rem]"
        >
          <p
            class="text-red-600 text-sm"
            id="error-radiatorLengthFrom"
          />
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <label
          class="text-sm"
          for="radiatorLengthTo"
        >
          Radiator length to (mm)
        </label>
        <input
          aria-errormessage="error-radiatorLengthTo"
          aria-invalid="false"
          class="p-2 border-2 h-10 bg-white"
          id="radiatorLengthTo"
          name="radiatorLengthTo"
          type="number"
        />
        <div
          class="h-[3rem]"
        >
          <p
            class="text-red-600 text-sm"
            id="error-radiatorLengthTo"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end gap-4"
    >
      <button
        class="border-2 rounded-lg px-10 py-3 text-sm font-bold cursor-pointer text-black bg-gray-100 hover:bg-gray-200"
        type="button"
      >
        Clear
      </button>
      <button
        aria-label="Search for radiators"
        class="border-2 rounded-lg px-10 py-3 text-sm font-bold cursor-pointer text-white bg-blue-700 hover:bg-blue-800"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</DocumentFragment>
`;
